### PR TITLE
Refactoring master branch to work with choreo implementation

### PR DESCRIPTION
**Pull request by chamilaadhi**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIConsumer.java [hunk 1]
```diff
@@ -241,7 +241,7 @@ void cleanUpApplicationRegistration(String applicationName, String tokenType, St
      * @throws APIManagementException
      */
     Set<SubscribedAPI> getPaginatedSubscribedAPIs(Subscriber subscriber, String applicationName, int startSubIndex,
-                                                  int endSubIndex, String groupingId) throws APIManagementException;
+                                                  int endSubIndex, String groupingId, String organization) throws APIManagementException;
```
Comment (bhathiya) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIConsumer.java:244]: We need this in  the javadoc comment. Please check all such cases.

Comment (Shakila) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/APIConsumer.java:246]: This is fixed now.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 2]
```diff
@@ -4637,24 +4662,31 @@ public int getApplicationsCount(int tenantId, String searchOwner, String searchA
...
         Connection connection = null;
         PreparedStatement prepStmt = null;
         ResultSet rs = null;
 
+        String uuid;
+        if (identifier.getUUID() != null) {
+            uuid = identifier.getUUID();
+        } else {
+            uuid = getUUIDFromIdentifier(identifier, organization);
+        }
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:4682]: these kind of branching can cause issues..
Shall we avoid passing APIIdentifier in the first place and pass uuid instead?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 3]
```diff
@@ -5392,15 +5430,126 @@ private int addSubscription(Connection connection, ApiTypeWrapper apiTypeWrapper
...
+     */
+    public List<String> getAPIVersionsMatchingApiNameAndOrganization(String apiName, String organization)
+            throws APIManagementException {
+
+        List<String> versionList = new ArrayList<String>();
+        try (Connection connection = APIMgtDBUtil.getConnection();
+             PreparedStatement ps = connection
+                     .prepareStatement(SQLConstants.GET_VERSIONS_MATCHES_API_NAME_AND_ORGANIZATION_SQL)) {
+            boolean initialAutoCommit = connection.getAutoCommit();
+            ResultSet resultSet = null;
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:5449]: use try with resources for PreparedStatement and ResultSet.

Please do this in all places below.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 4]
```diff
@@ -7239,25 +7386,22 @@ private void handleException(String msg, Throwable t) throws APIManagementExcept
...
         Map<Integer, URITemplate> uriTemplates = new LinkedHashMap<>();
         Map<Integer, Set<String>> scopeToURITemplateId = new HashMap<>();
         try (Connection conn = APIMgtDBUtil.getConnection();
              PreparedStatement ps =
                      conn.prepareStatement(SQLConstants.GET_URL_TEMPLATES_OF_API_WITH_PRODUCT_MAPPINGS_SQL)) {
-            ps.setString(1, APIUtil.replaceEmailDomainBack(identifier.getProviderName()));
-            ps.setString(2, identifier.getName());
-            ps.setString(3, identifier.getVersion());
+            ps.setString(1, uuid);
             try (ResultSet rs = ps.executeQuery()) {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:7405]: This is the ideal way (get rid of APIIdentifier and use uuid directly). We should try to do the same in other places too.
Let's try to do it in a separate round.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 5]
```diff
@@ -7990,6 +8135,38 @@ public APIProductIdentifier getAPIProductIdentifierFromUUID(String uuid) throws
...
+    public String getOrganizationByAPIUUID(String apiId) throws APIManagementException {
+        String organization = null;
+        try (Connection connection = APIMgtDBUtil.getConnection();
+             PreparedStatement ps = connection.prepareStatement(SQLConstants.GET_ORGANIZATION_BY_API_ID)) {
+            boolean initialAutoCommit = connection.getAutoCommit();
+            ResultSet result = null;
+            try {
+                connection.setAutoCommit(false);
+                ps.setString(1, apiId);
+                result = ps.executeQuery();
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8152]: PreparedStatement and ResultSet also should follow try with resources

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 6]
```diff
@@ -8018,24 +8195,57 @@ public String getUUIDFromIdentifier(APIIdentifier identifier) throws APIManageme
...
+                    uuid = resultSet.getString(1);
+                }
+            }
+        } catch (SQLException e) {
+            handleException("Failed to get the UUID for API : " + identifier.getApiName() + '-'
+                    + identifier.getVersion(), e);
+        }
+        return uuid;
+    }
+
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8227]: For now it is ok as a workaround. But We should try to get rid of getUUIDFromIdentifier type of methods in the future.
Let's avoid using this as much as possible. We should get to the root of the problem and get rid of APIIdentifier from the entire flow where we have to use this method.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 7]
```diff
@@ -8048,21 +8258,29 @@ public String getUUIDFromIdentifier(String provider, String apiName, String vers
...
-        try (Connection connection = APIMgtDBUtil.getConnection()) {
-            PreparedStatement prepStmt = connection.prepareStatement(sql);
+        PreparedStatement prepStmt = null;
+        String sql = SQLConstants.GET_UUID_BY_IDENTIFIER_AND_ORGANIZATION_SQL;
+        try {
+            if (connection == null) {
+                connection = APIMgtDBUtil.getConnection();
+                isNewConnection = true;
+            }
+            prepStmt = connection.prepareStatement(sql);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8279]: Let's never follow this approach. We may not expect developers to read the implementation of the function to identify they can pass null into 'connection'. 
Also `connection` is a very sensitive parameter so we must make sure it is close properly. Having conditional things like `isNewConnection` can prone for bugs.

Lets use a new overloaded function which doesn't have `Connection connection` in the signature and make only that new method 'public'. This method which has `Connection connection` should only be called inside current class hence let's make it private.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 8]
```diff
@@ -14936,6 +15191,41 @@ public void updateCategory(APICategory apiCategory) throws APIManagementExceptio
...
+             PreparedStatement statement =
+                     connection.prepareStatement(SQLConstants.GET_CATEGORIES_BY_ORGANIZATION_SQL)) {
+            statement.setString(1, organization);
+
+            ResultSet rs = statement.executeQuery();
+            while (rs.next()) {
+                String id = rs.getString("UUID");
+                String name = rs.getString("NAME");
+                String description = rs.getString("DESCRIPTION");
+                int tenantID = rs.getInt("TENANT_ID");
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:15213]: Can we get rid of TENANT_ID from this table?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 9]
```diff
@@ -207,6 +207,25 @@ private void updateGatewayPublishedAPIArtifacts(Connection connection, String ap
         }
     }
 
+    public String retrieveOrganization(String uuid) throws APIManagementException {
+
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:211]: Is this an API's uuid? Shall we add a method comment too?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java [hunk 10]
```diff
@@ -891,23 +893,33 @@ public static API prepareToCreateAPIByDTO(APIDTO body, APIProvider apiProvider,
         }
 
         //Get all existing versions of  api been adding
-        List<String> apiVersions = apiProvider.getApiVersionsMatchingApiName(body.getName(), username);
+        List<String> apiVersions = apiProvider.getApiVersionsMatchingApiNameAndOrganization(body.getName(),
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:896]: Hope this method addresses both tenant and choreo cases.
Can we remove getApiVersionsMatchingApiName method too?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java [hunk 11]
```diff
@@ -891,23 +893,33 @@ public static API prepareToCreateAPIByDTO(APIDTO body, APIProvider apiProvider,
...
+                                "Error occurred while adding API. API with name " + body.getName()
+                                        + " already exists with different context" + context  + " in the organization" +
+                                        " : " + organization,  ExceptionCodes.API_ALREADY_EXISTS);
+                    }
                 }
             }
         } else {
             //If no any previous version exists
-            if (apiProvider.isDuplicateContextTemplate(context)) {
+            if (apiProvider.isDuplicateContextTemplateMatchingOrganization(context, organization)) {
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:918]: Same as above

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java [hunk 12]
```diff
@@ -248,8 +247,8 @@ public Response getAllAPIs(Integer limit, Integer offset, String xWSO2Tenant, St
 
             APIProvider apiProvider = RestApiCommonUtil.getLoggedInUserProvider();
 
-            String tenantDomain = RestApiCommonUtil.getLoggedInUserTenantDomain();
-            boolean migrationMode = Boolean.getBoolean(RestApiConstants.MIGRATION_MODE);
+            String organization = (String) messageContext.get(RestApiConstants.ORGANIZATION);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java:250]: can we have a util method for this?

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java [hunk 13]
```diff
@@ -129,7 +134,8 @@ public Response applicationsGet(String groupId, String query, String sortBy, Str
         ApplicationListDTO applicationListDTO = new ApplicationListDTO();
 
         String username = RestApiCommonUtil.getLoggedInUsername();
-        
+        String organization = getOrganization(messageContext);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java:137]: I think we can make this method a utility method in a common class and use in other places too.

**Comment by chamilaadhi** [comment]
Integration Test results

Screenshot 2021-06-04 at 5 08 17 PM

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql [hunk 14]
```diff
@@ -1967,7 +1969,8 @@ CREATE TABLE IF NOT EXISTS AM_API_CATEGORIES (
   NAME VARCHAR(255),
   DESCRIPTION VARCHAR(1024),
   TENANT_ID INTEGER,
-  UNIQUE (NAME,TENANT_ID),
+  ORGANIZATION VARCHAR(100),
+  UNIQUE (NAME,TENANT_ID,ORGANIZATION),
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql:1973]: Hope we pass tenant domain properly in onPrem case. Hence do we need tenantID in the UNIQUE statement?
Also if we have properly ORGANIZATION populated from the code and the migration, we can drop TENANT_ID column from this.

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql [hunk 15]
```diff
@@ -1420,9 +1420,10 @@ CREATE TABLE IF NOT EXISTS AM_APPLICATION (
     UPDATED_TIME TIMESTAMP,
     UUID VARCHAR(256),
     TOKEN_TYPE VARCHAR(10),
+    ORGANIZATION VARCHAR(100),
```
Comment (malinthaprasan) [inline features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql:1423]: Shall we make the size consistent? 100 vs 256

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java (hunk 13 above)
Comment (chamilaadhi) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/impl/ApplicationsApiServiceImpl.java]: fixed from 0e19391

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java (hunk 12 above)
Comment (chamilaadhi) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApisApiServiceImpl.java]: fixed from 0e19391

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql (hunk 14 above)
Comment (senthuran16) [reply features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql:1973]: Fixed in https://github.com/wso2/carbon-apimgt/pull/10585/commits/37158abec84c566087fdcf7e29c9fd27849a9f6c

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 8 above)
Comment (senthuran16) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:15213]: Fixed in https://github.com/wso2/carbon-apimgt/pull/10585/commits/37158abec84c566087fdcf7e29c9fd27849a9f6c

File: features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql (hunk 15 above)
Comment (senthuran16) [reply features/apimgt/org.wso2.carbon.apimgt.core.feature/src/main/resources/sql/h2.sql:1423]: Fixed in https://github.com/wso2/carbon-apimgt/pull/10585/commits/00b44d406db58adcab547df7bcb783088089fbb5

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 7 above)
Comment (chamilaadhi) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8238]: @tharikaGitHub

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 6 above)
Comment (chamilaadhi) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: one place that this method is used is in API import util where we check whether the API exists. I think we may not be able to totally remove this method in a place like that.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 2 above)
Comment (chamilaadhi) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:4682]: @tharikaGitHub seems like there are no usages of getConsumerKeys() method.

Comment (tharikaGitHub) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:4682]: Will change this.

Comment (tharikaGitHub) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:4682]: This method is removed with https://github.com/wso2/carbon-apimgt/pull/10587

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 7 above)
Comment (tharikaGitHub) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:8238]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10587

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java (hunk 10 above)
Comment (tharikaGitHub) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:896]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10587

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java (hunk 11 above)
Comment (tharikaGitHub) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/PublisherCommonUtils.java:918]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10587

**Review by malinthaprasan — APPROVED** [review]

**Pull request merged**